README.md
Uncap
=====

Map Caps Lock key to Escape key, or any to any key, on Windows systems.

[![Download][Download SVG]][Download URL]
[![MIT License][License Badge]][L]
[![Twitter][Twitter SVG]][Twitter URL]

Apart from documenting how Uncap works on Windows, this README contains
detailed documentation about how to remap keys on Linux and Mac as well.

[Download SVG]: https://img.shields.io/badge/download-uncap.exe-brightgreen.svg
[Download URL]: https://github.com/susam/uncap/releases/download/0.3.0/uncap.exe
[License Badge]: https://img.shields.io/badge/license-MIT-blue.svg
[Twitter SVG]: https://img.shields.io/badge/twitter-%40susam-%231da1f2
[Twitter URL]: https://twitter.com/susam


Contents
--------

* [Features](#features)
* [Get Started](#get-started)
* [Custom Mapping](#custom-mapping)
* [Usage](#usage)
* [Known Issue with Microsoft Edge on Windows 10](#known-issue-with-microsoft-edge-on-windows-10)
* [Alternatives](#alternatives)
  * [Windows: Scancode Map Registry Value](#windows-scancode-map-registry-value)
  * [Windows: AutoHotkey](#windows-autohotkey)
  * [Linux: setxkbmap](#linux-setxkbmap)
  * [Linux: xmodmap](#linux-xmodmap)
  * [Linux: loadkeys](#linux-loadkeys)
  * [Mac: System Preferences](#mac-system-preferences)
  * [Mac: Seil](#mac-seil)
* [Resources](#resources)
* [License](#license)
* [Support](#support)


Features
--------

The following list briefly describes some of the features of Uncap.

 1. Map Caps Lock key to Escape key on a new Windows system with almost
    no effort. Just download and double-click. This is the primary
    reason why Uncap was written. This tool was written for users of
    vi/Vim editor who like to map their Caps Lock key to Escape key for
    convenience.
 2. Map any key to any key, multiple keys to other keys, or disable
    keys using command line arguments. No additional file or script is
    required.
 3. Enable, disable or modify key mappings without having to reboot the
    Windows system.
 4. Disable key mappings easily by stopping Uncap.
 5. It is a single-file executable. It is very lightweight. The
    executable is only about 100 KB in size. It occupies less than 1 MB
    of memory and negligible CPU while running.

There are several other methods and tools available on Windows to map
one key to another. But none of them seem to have all five features
enumerated above. For example, editing [Scancode
Map](#windows-scancode-map-registry-value) registry value requires the
Windows system to be rebooted after every change. On the other hand
[AutoHotkey](#windows-autohotkey) requires an additional script file to
be written. Therefore, Uncap was written to support all five features
described above.


Get Started
------------

Uncap is a single-file executable: [`uncap.exe`][Download URL].

The simplest way to run it is to download it and double-click it. No
output window is displayed. It runs in background. Once it is running,
whenever Caps Lock key is pressed, it is translated to the Escape key by
Uncap.


Custom Mapping
--------------

It is possible to override the default mapping of Caps Lock key to
Escape key by specifying a custom mapping, however to do so, Uncap must
be run with arguments either from Command Prompt or from Windows Run
dialog box. The following steps show how uncap.exe may be used to map
Caps Lock key to Left Control key.

 1. Copy `uncap.exe` to a directory specified in the Windows `PATH`
    variable. `C:\Windows` is a very convenient location to copy this
    file to.

 2. Launch Windows Run dialog box by pressing `Windows Logo Key` + `R`.

 3. Enter the following command.

        uncap 0x14:0xa2

 4. Press OK button. This would launch Uncap. No output window would be
    displayed. Uncap would run in background. Whenever Caps Lock key is
    pressed now, Uncap would translate it to Left Control key.

 5. To terminate Uncap, run the following command.

        uncap -k

In the above steps, Caps Lock key is mapped to Left Control key by
specifying `0x14:0xa2` as an argument to Uncap. The virtual-key code of
Caps Lock is `0x14` and the virtual-key code of Control key is `0xa2`.
Therefore, the argument `0x14:0xa2` maps Caps Lock key to Control key.

The list of virtual-key codes for every key can be found at
<https://msdn.microsoft.com/library/windows/desktop/dd375731.aspx>.


Usage
-----

When Uncap is run without any arguments, it maps Caps Lock key to Escape
key. It may be run with arguments to map Caps Lock key to any key, any
key to any key, or disable any key.

Each argument is a colon separated pair of virtual-key codes from
<https://msdn.microsoft.com/library/windows/desktop/dd375731.aspx>.

The virtual-key code may be specified either as hexadecimal integer as
mentioned in the above URL or its equivalent decimal notation. These
details are illustrated in the list of examples below.

 1. Map Caps Lock key to Escape key.

        uncap

 2. Swap Caps Lock key with Escape key.

        uncap 0x1b:0x14

    Caps Lock key is already mapped to Escape key by default. The above
    command maps Escape key to Caps Lock key, thus effectively swapping
    both keys.

 3. The hexadecimal integer is case-insensitive, so the following
    command is equivalent to the previous command.

        uncap 0x1B:0x14

 4. The virtual-key codes may be specified in decimal notation, so the
    following command is equivalent to the previous command.

        uncap 27:20

 5. Map Caps Lock key to Left Control key.

        uncap 0x14:0xa2

 6. Swap Caps Lock key and Left Control key.

        uncap 0x14:0xa2 0xa2:0x14

    This example shows that it is possible to specify more than one
    argument to map multiple keys to other keys.

 7. Here is another example that maps multiple keys to other keys. The
    following command maps Caps Lock key to Left Control key, Left
    Control key to Escape key and Escape key to Caps Lock key.

        uncap 0x14:0xa2 0xa2:0x1b 0x1b:0x14

 8. If a key is mapped to itself, then no mapping occurs for it. This
    may be used to override default behaviour of mapping Caps Lock key
    to Escape key and leave it unmapped while mapping another key to
    some key. Here is an example that unmaps Caps Lock key and maps F1
    key to escape key.

        uncap 0x14:0x14 0x70:0x1b

    When a key is mapped to another key, and that key is pressed, Uncap
    translates it to the key it is mapped to by synthesizing a new
    keystroke and injecting it into the system. But when a key is
    mapped to itself, and that key is pressed, it is ignored, thus no
    translation or synthesis of a new keystroke occurs.

 9. If a key is mapped to 0, then the key is disabled. Here is an
    example that disables Caps Lock key completely.

        uncap 0x14:0

10. Uncap runs without displaying a console by default. To terminate
    another instance of Uncap that may be running, use the `-k` or
    `--kill` option.

        uncap -k

    The above command kills all other instances of Uncap that are
    running with or without a console.

11. To run Uncap with a console, use the `-c` or `--console` option.

        uncap -c

    When Uncap is running in a console like this, it may be terminated
    by pressing `Ctrl` + `C`.

12. To run Uncap in debug mode, use the `-d` or `--debug` option.

        uncap -d

    The debug output contains one line of output with details about
    every press of a key or release of a key.

13. To log the keystroke details to a file, use the `-f` or `--file`
    option.

        uncap -f C:\keys.txt

    With this option alone, Uncap runs in background and logs the
    keystroke details to the specified file. This option may be
    combined with `-c` or `-d` to run Uncap in a console or print
    debug output on the console, respectively.

14. See the complete usage details along with options supported.

        uncap -h


Known Issue with Microsoft Edge on Windows 10
---------------------------------------------

When Uncap is downloaded using the Microsoft Edge web browser on Windows
10, it sometimes fails to map keys.

The following measures have found to be *unsuccessful* in resolving the
issue:

  - Right clicking `uncap.exe`, then selecting *Properties*, then
    selecting *Unblock*, and then clicking *OK* does not resolve the
    issue.

  - Right clicking `uncap.exe`, then selecting *Properties*, then going
    to *Security* tab, then clicking *Edit*, and then removing the
    following two accounts does not resolve the issue:

        Unknown Account(S-1-15-2-3624051433-2125758914-1423191267-1740899205-1073925389-3782572162-737981194)
        Unknown Account(S-1-15-3-3624051433-2125758914-1423191267-1740899205-1073925389-3782572162-737981194)

    These accounts are added automatically to files downloaded with
    Microsoft Edge.

The following workarounds have been found to resolve the issue
successfully:

  - Copy `uncap.exe` to a new file in the same directory. The new file
    when run remaps keys successfully. Finally, delete the downloaded
    `uncap.exe`, then rename the new file to `uncap.exe`, and use it
    normally.

  - Copy `uncap.exe` to some other directory. The new copied file when
    run remaps keys successfully.

This issue was observed with Microsoft Edge Microsoft Edge 44.17763.1.0
on Windows 10 Pro.


Alternatives
------------

There are other tools and methods available to map one key to another on
Windows, Linux and Mac. This section describes a few such tools and methods.


### Windows: Scancode Map registry value ###

One way to map a key to another key in Windows without using any
additional software is by editing the registry to add a value called
"Scancode Map" in "HKLM\SYSTEM\CurrentControlSet\Control\Keyboard
Layout" registry key.

For example, to map Caps Lock key to Escape key, create a registration
entries file, i.e. a file with reg as its extension, say uncap.reg, with
the following content.

    REGEDIT4
    [HKEY_LOCAL_MACHINE\SYSTEM\CurrentControlSet\Control\Keyboard Layout]
    "Scancode Map"=hex:00,00,00,00,00,00,00,00,02,00,00,00,01,00,3a,00,00,00,00,00

Then save the file. Then double-click this file to add it to Windows
registry. Then reboot the Windows system. After rebooting, whenever Caps
Lock key is pressed, it would function like Escape key.

The same registry value may be added by running the following command in
Command Prompt.

    reg add "HKLM\SYSTEM\CurrentControlSet\Control\Keyboard Layout" /v "Scancode Map" /t REG_BINARY /d 00000000000000000200000001003a0000000000

The hexadecimal code in the registry value is the scan code map. The
format of the scan code map is described in detail at
<https://msdn.microsoft.com/library/windows/hardware/jj128267.aspx#scan_code_mapper_for_keyboards>.

The following command may be used to view this registry value.

    reg query "HKLM\SYSTEM\CurrentControlSet\Control\Keyboard Layout" /v "Scancode Map"

The following command may be used to remove this registry value.

    reg delete "HKLM\SYSTEM\CurrentControlSet\Control\Keyboard Layout" /v "Scancode Map"

While this method has an advantage that it requires no additional
software, a disadvantage of this method is that it requires the system
to be rebooted after every change in the registry.

There is a tool called [SharpKeys][] that comes with graphical user
interface to manage this registry value. However a reboot is still
required after every change made with SharpKeys.

[SharpKeys]: https://sharpkeys.codeplex.com/


### Windows: AutoHotkey ###

There is a fairly sophisticated scripting language for Windows called
[AutoHotkey][] that can be used to map one key to another, among many
other things.

It is quite simple to map one key to another using AutoHotkey. For
example, after installing AutoHotkey, to map Caps Lock key to Escape
key, create an AutoHotkey script, i.e. a file with ahk as its extension,
say uncap.ahk, with the following content.

    CapsLock::Esc

Then save the file. Then double-click this file to run AutoHotkey with
this script. Once this is done, whenever Caps Lock key is pressed, it is
translated to Escape key. An AutoHotkey icon appears in the notification
area (system tray). To stop it, right click on the icon and select
'Exit'.

This method has an advantage that it does not require the system to be
rebooted. However, AutoHotkey is a scripting language that can be used
to do a number of other things such as defining hotkeys (shortcut keys),
hotstrings (auto-replace), macros (scripted actions), etc. This may feel
akin to killing a fly with a sledgehammer if one just wants to map a key
to another key.

[AutoHotkey]: http://www.autohotkey.com/


### Linux: setxkbmap ###

On Linux, the `setxkbmap` command may be used to map one key to another
key in the X Window System. For example, the following command maps Caps
Lock key to Escape key.

    setxkbmap -option caps:escape

The list of options along with a short description for each can be found

CHANGES.md
Changelog
=========

0.4.0 (UNRELEASED)
------------------

### Changed

- Concise debug log format, so that it is easier to read.


0.3.0 (2019-03-24)
------------------

### Added 

- Support mapping of injected keyboard events.
- Show `dwExtraInfo`, `nCode`, and `flags` in debug output.
- Show `LLKHF_LOWER_IL_INJECTED` flag as `LOW` in debug output.

### Fixed

- Issue that prevents Uncap console window from becoming hidden when
  Cmder is running. With this release, Uncap console window disappears
  as expected regardless of whether Cmder is runing or not.


0.2.2 (2015-02-15)
------------------

### Fixed

- Possible buffer overflows while handling command line arguments.


0.2.1 (2015-01-15)
------------------

### Changed

- Minor changes in the help and version texts.


0.2.0 (2015-12-22)
------------------

### Added

- Write verbose logs to file with `-f` or `--file` option.

### Changed

- Switch from BSD 2-Clause License to MIT License.


0.1.0 (2015-09-13)
------------------

### Added

- Map Caps Lock key to Escape key by default.
- Map any key to any key with `MAP_KEY:TO_KEY` argument.
- No mapping occurs with `MAP_KEY:MAP_KEY` argument.
- Disable any key with `MAP_KEY:0` argument.
- Run in background without console by default.
- Kill other instances of the program with `-k` or `--kill` option.
- Run silently in console with `-c` or `--console` option.
- Run verbosely in console with `-d` or `--debug` option.


uncap.c
/**
@file
@mainpage

Map Caps Lock key to Escape key, or any key to any key.

The MIT License (MIT)
---------------------
Copyright (c) 2015-2021 Susam Pal

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
"Software"), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
*/


#include <stdio.h>
#include <ctype.h>
#include <windows.h>
#include <tlhelp32.h>


/** Version of the program. */
#define VERSION "0.4.0-dev"

/** Author of the program. */
#define AUTHOR "Susam Pal"

/** Copyright notice. */
#define COPYRIGHT "Copyright (c) 2015-2021 " AUTHOR

/** URL to a copy of the license. */
#define LICENSE_URL "<https://susam.github.io/licenses/mit.html>"

/** URL to report issues to. */
#define SUPPORT_URL "<https://github.com/susam/uncap/issues>"

/** Maximum length of command line argument that is read. */
#define MAX_ARG_LEN 16

/** Maximum allowed length of an error message. */
#define MAX_ERR_LEN 256

/** Magic number to identify keyboard input injected by Uncap. */
#define UNCAP_INFO (WM_APP + 3195) /* 35963 */

#ifndef LLKHF_LOWER_IL_INJECTED
/** Workaround for missing definition in MinGW winuser.h. **/
#define LLKHF_LOWER_IL_INJECTED 0x00000002
#endif

/**
Check if two null-terminated byte strings are equal.

@param a Pointer to null-terminated byte string. (type: const char *)
@param b Pointer to null-terminated byte string. (type: const char *)

@return 1 if the two strings are equal; 0 otherwise.
*/
#define streq(a, b) (strcmp(a, b) == 0)


/**
Copy null-terminated byte string into a character array.

@param a Pointer to character array to copy the string to. (type: char *)
@param b Pointer to null-terminated byte string to copy. (type: const char *)
@param c Maximum number of characters to copy. (type: size_t)

@return a
*/
#define strcp(a, b, c) (a[0] = '\0', strncat(a, b, c - 1))


/**
Convert all characters of a string to lowercase.

@param s Pointer to null-terminated byte string.

@return s
*/
char *strlower(char *s)
{
    int i;
    for (i = 0; s[i] != '\0'; i++)
        s[i] = (char) tolower(s[i]);
    return s;
}


/**
Return name of the leaf directory or file in the specified path.

Both backslash and forward slash are treated as path separators. A
pointer to the beginning of the substring between the last slash
(exclusive) and the end of the string is returned as the basename.
Therefore if the specified path ends with a slash then an empty string
is returned.

@param path Path string.

@return Name of the leaf directory or file in the specified path.
*/
const char *basename(const char *path)
{
    const char *base;
    if ((base = strrchr(path, '\\')) != NULL)
        return base + 1;
    else if ((base = strrchr(path, '/')) != NULL)
        return base + 1;
    else
        return path;
}


/**
Values returned by a function to indicate success or failure.

These return codes may be returned by a function to indicate success or
failure of its operation as well as the next course of action.
*/
enum action {
    GOOD, /**< Successful operation; program should continue. */
    EXIT, /**< Successful operation; program should exit normally. */
    FAIL  /**< Failed operation; program should exit with error. */
};


/**
Global state of this program.
*/
struct state {
    char name[MAX_ARG_LEN];  /**< Program name. */
    WORD keymap[256];        /**< Key mappings. */
    HHOOK hook;              /**< Handle to keyboard hook procedure. */
    int console;             /**< Whether console mode is enabled. */
    int debug;               /**< Whether verbose mode is enabled. */
    FILE *file;              /**< File to write verbose logs to. */
    char error[MAX_ERR_LEN]; /**< Error message for failed operation. */
} my; /**< Global state of this program. */


/**
Output error message on the standard error stream.

@param format Format string for printf.
@param ...    Additional arguments.

@return EXIT_FAILURE; the caller of this function may return this code
        to indicate abnormal termination of the program.
*/
int error(const char *format, ...)
{
    va_list ap;
    va_start(ap, format);
    fprintf(stderr, "%s: ", my.name);
    vfprintf(stderr, format, ap);
    return EXIT_FAILURE;
}


/** Log format string. */
#define LOG_FMT \
    "%-10s %3d %5lu %3lu " \
    "%3lu (%#04lx) %3lu (%#04lx) " \
    "[%s%s%s%s%s%s%s]\n"


/**
Log details of a key stroke to a specified file.

@param file File to write to. (type: FILE *)
*/
#define logKeyTo(file) \
            fprintf(file, LOG_FMT, \
                    wParamStr, nCode, p->dwExtraInfo, p->flags, \
                    p->scanCode, p->scanCode, p->vkCode, p->vkCode, \
                    vkStr, upStr, extStr, altStr, lowStr, injStr, uncapStr)

/**
Log details of a key stroke.

@param nCode  Code used to determine how to process the message.
@param wParam Identifier of the keyboard message.
@param lParam Pointer to KBDLLHOOKSTRUCT structure.
*/
void logKey(int nCode, WPARAM wParam, LPARAM lParam)
{
    const KBDLLHOOKSTRUCT *p = (KBDLLHOOKSTRUCT *) lParam;

    char wParamStr[16];
    char vkStr[16];
    char extStr[16];
    char lowStr[16];
    char injStr[16];
    char altStr[16];
    char upStr[16];
    char uncapStr[16];

    /* Translate identifier of keyboard message to string notation. */
    switch (wParam) {
    case WM_KEYDOWN:
        strcpy(wParamStr, "KEYDOWN");
        break;

    case WM_KEYUP:
        strcpy(wParamStr, "KEYUP");
        break;

    case WM_SYSKEYDOWN:
        strcpy(wParamStr, "SYSKEYDOWN");
        break;

    case WM_SYSKEYUP:
        strcpy(wParamStr, "SYSKEYUP");
        break;
    }

    /* Translate virtual-key code to string notation. */
    if (p->vkCode == VK_RETURN)
        sprintf(vkStr, "RETURN");
    else if (p->vkCode == VK_CAPITAL)
        sprintf(vkStr, "CAPITAL");
    else if (p->vkCode == VK_ESCAPE)
        sprintf(vkStr, "ESCAPE");
    else if (p->vkCode == VK_LCONTROL)
        sprintf(vkStr, "LCONTROL");
    else if (p->vkCode == VK_RCONTROL)
        sprintf(vkStr, "RCONTROL");
    else if (p->vkCode == VK_LMENU)
        sprintf(vkStr, "LMENU");
    else if (p->vkCode == VK_RMENU)
        sprintf(vkStr, "RMENU");
    else if (p->vkCode == VK_LWIN)
        sprintf(vkStr, "LWIN");
    else if (p->vkCode == VK_RWIN)
        sprintf(vkStr, "RWIN");
    else if ((p->vkCode >= '0' && p->vkCode <= '9') ||
             (p->vkCode >= 'A' && p->vkCode <= 'Z'))
        sprintf(vkStr, "%c", p->vkCode);
    else if (p->vkCode >= VK_NUMPAD0 && p->vkCode <= VK_NUMPAD9)
        sprintf(vkStr, "NUMPAD%d", p->vkCode - VK_NUMPAD0);
    else if (p->vkCode >= VK_F1 && p->vkCode <= VK_F24)
        sprintf(vkStr, "F%d", p->vkCode - VK_F1 + 1);
    else
        sprintf(vkStr, "%#x", p->vkCode);

    /* Translate each flag to string notation. */
    strcpy(extStr, p->flags & LLKHF_EXTENDED ? " EXT" : "");
    strcpy(lowStr, p->flags & LLKHF_LOWER_IL_INJECTED ? " LOW": "");
    strcpy(injStr, p->flags & LLKHF_INJECTED ? " INJ" : "");
    strcpy(altStr, p->flags & LLKHF_ALTDOWN ? " ALT" : "");
    strcpy(upStr, p->flags & LLKHF_UP ? " UP" : " DN");
    strcpy(uncapStr, p->dwExtraInfo == UNCAP_INFO ? " UNCAP" : "");

    /* Log key to standard error stream if verbose mode is enabled. */
    if (my.debug)
        logKeyTo(stderr);

    /* Log key to user specified file if file logging is enabled. */
    if (my.file != NULL)
        logKeyTo(my.file);

    fflush(NULL);
}


/**
Map one key to another key.

@param nCode  Code used to determine how to process the message.
@param wParam Identifier of the keyboard message.
@param lParam Pointer to KBDLLHOOKSTRUCT structure.

@return 1 if the keyboard message is processed, i.e. a key is mapped to
        another key, otherwise call CallNextHookEx and return the value
        it returns.
*/
LRESULT CALLBACK keyboardHook(int nCode, WPARAM wParam, LPARAM lParam)
{
    KBDLLHOOKSTRUCT *p = (KBDLLHOOKSTRUCT *) lParam;
    WORD keyCode = (WORD) p->vkCode;
    WORD mapCode = my.keymap[keyCode];

    if (my.debug || my.file) {
        logKey(nCode, wParam, lParam);
    }

    if (mapCode == 0) {
        /* If key pressed is unmapped, disable the key press. */
        return 1;
    } else if (keyCode != mapCode && p->dwExtraInfo != UNCAP_INFO &&
               nCode >= 0) {
        /* If key is mapped, translate it to what it is mapped to. */
        INPUT inputs[1];
        PKEYBDINPUT ki = &inputs[0].ki;

        inputs[0].type = INPUT_KEYBOARD;
        ki->time = ki->wScan = 0;

        ki->wVk = mapCode;
        ki->dwFlags = (wParam == WM_KEYUP || wParam == WM_SYSKEYUP)
                      ? KEYEVENTF_KEYUP : 0;
        ki->dwExtraInfo = UNCAP_INFO;

        SendInput(1, inputs, sizeof *inputs);
        return 1;
    }

    return CallNextHookEx(my.hook, nCode, wParam, lParam);
}


/**
Kill other running instances of this program.

@return Next action to take based on whether this function could
        terminate other running instances of this program or not. FAIL
        is returned if this function failed to terminate at least one
        other running instance of this program. EXIT is returned if this
        function successfully terminated all other running instances of
        this program or if there were no other running instances of this
        program.
*/
enum action kill(void)
{
    char myExeFile[MAX_ARG_LEN + 4];
    char *dot;
    PROCESSENTRY32 entry;
    HANDLE snapshotHandle;
    int failure = 0;

    /* Take a snapshot of all processes running on the system. */
    snapshotHandle = CreateToolhelp32Snapshot(TH32CS_SNAPPROCESS, 0);
    if (snapshotHandle == NULL) {
        sprintf(my.error, "Cannot take snapshot of processes; ",
                          "error: %lu.", GetLastError());
        return FAIL;
    }

    /* Begin iterating through each process in the snapshot. */
    entry.dwSize = sizeof entry;
    if (!Process32First(snapshotHandle, &entry)) {
        sprintf(my.error, "Cannot retrieve process from snapshot; ",
                          "error: %lu.", GetLastError());
        return FAIL;
    }

    /* Name of the process to be killed. */
    strcpy(myExeFile, my.name);
    strlower(myExeFile);
    dot = strrchr(myExeFile, '.');
    if (dot == NULL || !streq(dot, ".exe"))
        strcat(myExeFile, ".exe");

    /* Iterate through each process in the snapshot, find the process to
       be killed and kill it. */
    do {
        HANDLE processHandle;

        /* Ignore current process. */
        if (entry.th32ProcessID == GetCurrentProcessId())
            continue;

        /* Ignore other programs. */
        if (!streq(strlower(entry.szExeFile), myExeFile))
            continue;

        /* Open another instance of this program. */
        processHandle = OpenProcess(PROCESS_ALL_ACCESS, 0,
                                    entry.th32ProcessID);
        if (processHandle == NULL) {
            error("Cannot open process \"%s\" (PID %lu); error %lu.",
                  entry.szExeFile, entry.th32ProcessID, GetLastError());
            failure = 1;
            continue;
        }

        /* Terminate another instance of this program. */
        if (TerminateProcess(processHandle, 0)) {
            printf("Terminated %s (PID %lu).\n",
                   entry.szExeFile, entry.th32ProcessID, GetLastError());
        } else {
            error("Cannot terminate %s (PID %lu); error %lu.\n",
                  entry.szExeFile, entry.th32ProcessID, GetLastError());
            failure = 1;
        }
        CloseHandle(processHandle);

    } while (Process32Next(snapshotHandle, &entry));

    CloseHandle(snapshotHandle);

    if (failure) {
        sprintf(my.error, "Failed to terminate all running instances "
                          "of %s.", myExeFile);
        return FAIL;
    }

    return EXIT;
}


/**
Show usage and help details of this program.
*/
void showHelp(void)
{
    const char *usage =
"Usage: %s [-k] [-c] [-d] [-f FILE] [-h] [-v] [[MAP_KEY:TO_KEY]...]\n\n";

    const char *summary =
"Map Caps Lock key to Escape key, or any key to any key.\n\n";

    const char *description1 =
"Caps Lock key is mapped to Escape key by default. This may be\n"
"overridden by specifying a new mapping for Caps Lock key. Any key\n"
"may be mapped to any key with one or more MAP_KEY:TO_KEY arguments.\n"
"Each argument is a colon separated pair of virtual-key codes from\n"
"<https://msdn.microsoft.com/library/windows/desktop/dd375731.aspx>.\n\n"

"The virtual-key code may be specified either as hexadecimal integer\n"
"as mentioned in the above URL or its equivalent decimal notation.\n\n";

    const char *description2 =
"If MAP_KEY equals TO_KEY, then no mapping occurs for it. If TO_KEY\n"
"equals 0, then the key mapped to 0 is disabled.\n\n";

    const char *details =
"Options:\n"
"  -k, --kill       Kill other instances of uncap.\n"
"  -c, --console    Run silently in console.\n"
"  -d, --debug      Run verbosely in console.\n"
"  -f, --file FILE  Write verbose logs to file.\n"
"  -h, --help       Show this help and exit.\n"
"  -v, --version    Show version and exit.\n\n"

"Arguments:\n"
"  MAP_KEY          Virtual-key code of key to map.\n"
"  TO_KEY           Virtual-key code of key to map to.\n\n"

"Report bugs to " SUPPORT_URL ".\n";

    printf(usage, my.name);
    printf(summary);
    printf(description1);
    printf(description2);
    printf(details);
}


/**
Show version and copyright details of this program.
*/
void showVersion(void)
{
    char name[MAX_ARG_LEN];

    const char *s =
    "%s " VERSION "\n"
    COPYRIGHT "\n\n"

    "This is free and open source software. You can use, copy, modify,\n"
    "merge, publish, distribute, sublicense, and/or sell copies of it,\n"
    "under the terms of the MIT License. You can obtain a copy of the\n"
    "MIT License at " LICENSE_URL ".\n\n"

    "This software is provided \"AS IS\", WITHOUT WARRANTY OF ANY KIND,\n"
    "express or implied. See the MIT License for details.\n";

    strcpy(name, my.name);
    name[0] = (char) toupper(name[0]);
    printf(s, name);
}


/**
Output love.
*/
void qtpi(void)
{
    const int show[] = {'.', '@', ' ', '\n'};
    const int love[] = {
        252, 7, 252, 7, 72, 6, 29, 6, 36, 6, 29,
        6,  72, 7, 60, 6,  17, 22, 17, 6, 12, 6,

              17, 22,           17,  6,
           60, 7, 52, 6,     13,  46,  13,
         6, 13, 46, 13, 6, 52, 7, 48, 6, 13,
        62, 5, 62, 13,  6, 48,  7, 48, 6, 13,
        18, 1, 70, 13,  6, 48,  7, 48, 6, 13,
        126, 13, 6, 48, 7, 52, 6, 13,  38, 2,
         38, 13, 6,  52, 7,  60, 6,  17, 94,
           17, 6,  60, 7,  72, 6,  17, 22,
              3, 18,  17, 6, 72, 7, 84,
                 6, 17,  46,  17, 6,
                    84, 7, 96, 6,
                       17, 22,
                         17,

        6, 96, 7, 108, 6,  13, 6, 13, 6, 108, 7,
        120,  6, 5,  6, 120, 7,  252, 7, 252, 7,
    };
    const char *say[] = {"", "Cutie Pai,", "I love you!", "-- Susam"};
    size_t i;
    int j;
    for (i = 0; i < sizeof love / sizeof *love; i++) {
        if (love[i] < 4)
            printf(say[love[i]]);
        else
            for (j = 0; j < love[i] / 4; j++)
                putchar(show[love[i] % 4]);
    }
}


/**
Parse command line arguments.

@param argc Argument count
@param argv Argument vector

@return Next action to take based on whether parsing of command line
        arguments was successful or not. GOOD is returned if command
        line arguments were parsed successfully and the program should
        continue. EXIT is returned if command line arguments were parsed
        and processed successfully and the program should exit. FAIL is
        returned if an error was found in command line arguments.
*/
enum action parseArguments(int argc, const char **argv)
{
    int i;

    /* Set default values. */
    strcp(my.name, basename(argv[0]), sizeof my.name);

    for (i = 0; i < (int) (sizeof my.keymap / sizeof *my.keymap); i++)
        my.keymap[i] = (WORD) i;

    my.keymap[VK_CAPITAL] = VK_ESCAPE;

    my.console = 0;
    my.debug = 0;
    my.file = NULL;

    /* Parse command line options. */
    i = 1;
    while (i < argc) {
        const char *arg = argv[i];
        if (streq(arg, "-h") || streq(arg, "--help")) {
            showHelp();
            return EXIT;
        } else if (streq(arg, "-v") || streq(arg, "--version")) {
            showVersion();
            return EXIT;
        } else if (streq(arg, "--qtpi")) {
            qtpi();
            return EXIT;
        } else if (streq(arg, "-c") || streq(arg, "--console")) {
            my.console = 1;
            ++i;
        } else if (streq(arg, "-d") || streq(arg, "--debug")) {
            my.debug = 1;
            ++i;
        } else if (streq(arg, "-f") || streq(arg, "--file")) {
            if (i == argc - 1) {
                sprintf(my.error, "Option '%.*s' must be followed by "
                                  "file path", MAX_ARG_LEN, arg);
                return FAIL;
            }
            arg = argv[++i];
            if ((my.file = fopen(arg, "a")) == NULL) {
                sprintf(my.error, "Cannot open %.*s.", MAX_ARG_LEN, arg);
                return FAIL;
            }
            ++i;
        } else if (streq(arg, "-k") || streq(arg, "--kill")) {
            return kill();
        } else if (streq(arg, "--")) {
            ++i;
            break;
        } else if (arg[0] == '-' && arg[1] != 0) {
            sprintf(my.error, "Unknown option '%.*s'.", MAX_ARG_LEN, arg);
            return FAIL;
        } else {
            break;
        }
    }

    /* Parse command line options. */
    while (i < argc) {
        const char *arg = argv[i];
        char *colon;
        unsigned long int mapKey;
        unsigned long int toKey;

        if ((colon = strchr(arg, ':')) == NULL) {
            sprintf(my.error, "Colon is missing from argument '%.*s'.",
                    MAX_ARG_LEN, arg);
            return FAIL;
        }

        mapKey = strtoul(arg, NULL, 0);
        toKey = strtoul(colon + 1, NULL, 0);

        if (mapKey < 1 || mapKey > 254 || toKey > 254) {
            sprintf(my.error, "Invalid key code in argument '%.*s'.",
                    MAX_ARG_LEN, arg);
            return FAIL;
        }

        my.keymap[mapKey] = (WORD) toKey;
        ++i;
    }

    /* Command line arguments parsed successfully. */
    return GOOD;
}


/**
Start the program.

@param argc Argument count.
@param argv Argument vector.

@return EXIT_SUCCESS if the program terminates normally;
        EXIT_FAILURE if an error occurs.
*/
int main(int argc, char **argv)
{
    enum action a;
    MSG msg;

    /* Parse command line arguments. */
    if ((a = parseArguments(argc, (const char **) argv)) == FAIL)
        return error(my.error);
    else if (a == EXIT)
        return EXIT_SUCCESS;

    /* Set visibility of console. */
    if (!my.console && !my.debug) {
        HWND h = GetConsoleWindow();
        if (h != NULL)
            ShowWindow(h, SW_HIDE);
        else
            error("Cannot find console window; error %lu.", GetLastError());
    }

    /* Install hook to monitor low-level keyboard input events. */
    my.hook = SetWindowsHookEx(WH_KEYBOARD_LL, keyboardHook, NULL, 0);
    if (my.hook == NULL)
        return error("Cannot install hook; error %lu.", GetLastError());

    /* Begin message delivery. */
    if (GetMessage(&msg, NULL, 0, 0) == -1)
        return error("Cannot retrieve message; error %lu.", GetLastError());

    return EXIT_SUCCESS;
}
